src/app/services/article.service.ts

```diff
@@ -73,6 +79,51 @@ export class ArticleService {
     return this.articles;
   };
 
+  public fetchAllArticlesFromFirestore(): Observable<Article[]> {
+    const url = `${this.BASE_FIRESTORE}/projects/auxilium-420904/databases/aux-db/documents:runQuery`;
+    const headers: HttpHeaders = new HttpHeaders({'Content-Type': 'application/json'});
+    const body = {
+      structuredQuery: {
+        from: [{collectionId: 'articles'}],
+      }
+    };
+
+    return this.http.post(url, body, {headers}).pipe(
+      map((documents: any): Article[] => {
+
+        const fetchedArticles: Article[] = [];
+        documents.forEach((document: any) => {
+          const fields = document.document.fields;
+          let documentId: string = '';
+
+          /* Todo: This is bad. We need our database to adhere to the contract of data blobs that we expect.
+              There is not a single article in the firestore that contains all requirements of the Article interface.
+          */
+          const article: Article = {
+            header: fields?.header?.stringValue?.toString() || INVALID_ARTICLE.header,
+            body: fields?.body?.stringValue?.toString() || INVALID_ARTICLE.body,
+            imageURI: fields?.imageURI ? fields.imageURI.stringValue : this.defaultImageURI,
+            meta: {
+              name: document?.name?.toString(),
+              documentId,
+              category: fields?.meta?.mapValue?.fields?.category?.stringValue || INVALID_ARTICLE.meta?.category,
```

Comment:
`category` and `subCategory` should be optional. At the time the article is written, you might not know what categories to add... unless we want a category called "uncategorized"..

For sure, `subCategory` is optional. We don't have subCategories for "High School/College" and "Striving for Excellence" categories.

Reply:
When this project first started, this is exactly why I prosed taking about the data's structure of our elements.

We are now in a state of intertwining discussion about editing / uploading articles, when we haven't talked about and agreed on what an article actually represents. I now understand why you changed the Article Interface to optional fields previously as I went through a lot of "accessing undefined field" issues during my fetching of All articles from the firestore and trying to display them as Articles.

We should probably step back and define the following:

- What is an Article?
 - What fields are included?
 -  What is optional in an Article?

- Same questions about the new  _Meta_ interface.
 

After that is defined we can ask these questions:

- What can a user ( Admin ) edit on an Article?
- What can a user ( Admin ) upload ( New Article ) on an Article?
- When and how is _Meta_ generated after a creation / edit of an Article?

I'm not saying this is 100% set in stone definition, as that would be silly, as we're always going to update things.

I say this because there is no way that we should be attaching **_"?."_** to all of our queries. Not only is it messy, its highly unreliable. We need some sort of footing of confidence here, and if those _required_ fields do NOT exist, it's simply an _invalid-article_.

Reply:
When I say **_Article_** I'm asking about the our unique project's Class / Interface **_Article_** not the webster's definition of "article."

Our _**Article**_ can represent anything and have any number of fields, but those need to be discussed on what is _required_ and therefore always accessible on every _**Article**_ and what is not, i.e. _**optional**_.

Example:

Article Interface:

- header
- body
- ?headerImage
- articleColor
- ?articleQuotes
- ?likes
- ?comments
- Meta
- SubMeta
- TriMeta

Our _**Article**_ could represent anything and contain anything, so we need to articulate what an _**Article**_ is subject to PRs and our technical meetings.

Reply:
And I'm saying all of those are optional except for header and body.
If it's in firebase, there will always be meta/documentId and name.

Reply:
You still have an article even if you don't have any of those other things.

Reply:
> When I say Article I'm asking about the our unique project's Class / Interface Article not the webster's definition of "article."

I fully understand what you meant. My point is that that `Article` class should be model the article as understood by the reader and the author. That should guide us on what's required and not required.

From the technical perspective, you need IDs (index), so that `article-id` (like a page number) and `documentId` and `name` for accessing it from firestore.

We also have the timestamps. But these are all givens. I'm not sure why any of the other fields are required.

Please disregard all other fields not modeled in `article.ts`. I'll have those cleaned-up. There were from the original implementation.

Reply:
Well, currently not. This is what I meant by agreement on structures.

I built an Article with the following requirements:

- articleId
 - imageURI
 - body
- _**category**_
 -  meta?
 
 Category was then moved into Meta by you, which is a **_optional_** field, which would completely break my Articles if _**Meta**_ doesn't exist.

This was clearly a problem when you were fetching featured articles as that's why you changed all the requirements to "optional" so that you could get things working, but you committed those changes.

I changed them back, because that was what was agreed on, and a lot of work followed.

I'm just asking to discuss class / interface structure of the project so we are on the same page with what is guaranteed and what is not. Always subject to change in PRs.

Reply:
`category` should be optional.

Reply:
category is not part of what constitutes an "article". It's data about the article, so it should be under "meta".

Reply:
It seems that there's a way to make meta required in some cases
```
// Article where meta is required
export interface PreviewArticle extends Omit<Article, 'meta'> {
  meta: Required<ArticleMeta>;
}
```

Reply:
> category is not part of what constitutes an "article". It's data about the article, so it should be under "meta".

Ok, so this is a perfect example.

Whether or not this is true, everything I had built for `Article` was represented in a slightly different way **_without_** `meta` and `category` as one of the top-level members. This was discussed with you prior if it was ok, and we both agreed.

Then, not only was `meta` added, but, you moved a `required` member for my `PreviewCardComponents` (`category`) into `meta` which was then assigned as `optional`. This is significant, because if there is some reason `meta` does not exist on an Article, `category` will be empty / undefined, and the `PreviewCard` will have missing text on the bottom, and thus, `PreviewCards` would not look symmetrical. `category` being `optional` was not part of my original design for the custom html element, hence, `required`.

You had previously called me out for deleting unused code and asked me to ask you to make those modifications, I guess I'm asking the same courtesy here.

At the end of the day, I really have no issues renaming, reorganizing, adding new members / fields to any data type or structure. I wouldn't even mind getting rid of my work altogether if that's what we agreed on. But, all I ask is that we both agree and understand what's going on, otherwise, it's very difficult for developers to live on islands and then to self research what the other developers on other islands have done. That's why I proposed using PRs to have quick discussions about what's going on before merging.

Reply:
- It's not the same thing. I refactored the code. You deleted the entire method.

The feature was fully intact. You assumed that sub-category is  required but it's the wrong assumption. Category and subcategory should not be required globally. Maybe for featured articles, maybe it's required but there might still be a case when a featured article might not have a catetory.

I sense frustration on your end but I have a similar frustration when the article view started failing when the article is not a featured article. We're two different people from very different backgrounds. We're bound to have string disagreements.

Anyway, I'm glad you're raising these issue so we know what each other is thinking about.

Just an FYI, I needed to merge my changes because I wanted to make sure it worked when deployed.

I'll not have that problem anymore because I "forked" the repo into aux-ng-dev. https://everking.github.io/aux-ng-dev/home. 

In this new repo, I experimented GH actions job pulling all articles and committing it to git. Check it out if you're interested. https://github.com/everking/aux-ng-dev